### DISCUSSION TITLE
Smart Meter Interface/Scripting: Decode multiple string lines

### DISCUSSION THREAD

**Discussion by GolterU** [arendst/Tasmota#17434] [Q&A] [answered]
Hello,
I successfully read the meter ID from my Smart Meter ZPA ZE311 (OBIS-ASCII):
Raw data: `0.0.0(1ZPA5723035xyz)`
Script line: `1,0.0.0(@#),meter ID ,,Meter_ID,0`
Web UI display: `Electricity meter ID "1ZPA5723035xyz"`

However, the meter also supplies 25 lines of historical data (meter readings).
Raw data:
...
`1.8.0*09(0007797.306*kWh)`
`1.8.0*10(0007628.544*kWh)`
`1.8.0*11(0007418.247*kWh)`
...
and the corresponding Timestamps in separate lines:
...
`0.1.2*09(220301000000)`
`0.1.2*10(220201000000)`
`0.1.2*11(220101000000)`
...

The Timestamps are in the format `YYMMDDhhmmss`
A line with
`1,0.1.2*09(@#),time,,DateTime09,0`
to decode already works. But how do I do that with the other entries if **only one** is allowed per meter ([Meter Metrics](https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-metrics) at `scale`)???

The decode as a numeric value with
`1,0.1.2*09(@1,timestamp09:,,DateTime09,0`
obviously exceeds the range of values of a floating point variable. Numerical decoding always returns the number 2147483647 (2^31 - 1).
Is there a solution to decode multiple string lines?

Golter

**Comment by gemu2015** [comment, marked as answer]
all numbers in this driver are double precision because meter values can get really large. i however just realized that the function dtostrfd() passes a double but does a call to dtostrf  (float) thus reducing the resolution. i will modify the driver with a real double to string conversion and make a pr soon.

**Reply by GolterU** [reply]
Hello gemu,
thank you for your quick response and your efforts.
Golter

**Reply by gemu2015** [reply]
there will be no more than one string per meter.

could you please try this version with timestamp as number
it now decodes with double resolution

[xsns_53_sml.ino.zip](https://github.com/arendst/Tasmota/files/10256513/xsns_53_sml.ino.zip)

**Reply by gemu2015** [reply]
remark. since the timestamp is very large as a number you may need to set the divider to 10000

**Reply by GolterU** [reply]
Hello gemu,
I tested the modified file "xsns_53_sml.ino". The 10-digit decimal should fit comfortably in a double-precision variable. The division with the scale parameter (see below) is carried out, but apparently the original value has already been truncated. In any case, the result is still wrong, e.g.:
source data
...
`0.1.2*01(221101000000)`
or
`0.1.2*25(201101000000)`
...

decode line:
...
`1,0.1.2*25(@1,date oldest value:,,Date_oldest,0`
or
`1,0.1.2*01(@1000000,date youngest value:,,Date_youngest,0`
...

Results:
...
`Current date oldest value: 2147483647`
or
`Current date most recent: 2147`
...

Did I maybe do something wrong when compiling?
I'm using Gitpod's web-based compiler. I unzipped your file "[xsns_53_sml.ino.zip](https://github.com/arendst/Tasmota/files/10256513/xsns_53_sml.ino.zip)" and copied the content into the file of the same name "/workspace/Tasmota/tasmota/tasmota_xsns_sensor/xsns_53_sml.ino" and compiled. The file is left in the "modified" state. The Tasmota master was used and later also the development branch, each with the adapted "user_config_override.h".

What else can I do?

Thanks for further help!
Golter

**Reply by gemu2015** [reply]
ok, found the bug and fixed

[xsns_53_sml.ino.zip](https://github.com/arendst/Tasmota/files/10295769/xsns_53_sml.ino.zip)

**Reply by GolterU** [reply]
Hi gemu,
**many thanks for the support! It now works as intended.**

What I noticed (but doesn't bother me):
If the raw data line
`0.1.2*01(221101000000)`
is decoded as a complete 12-digit decimal number with
`1,0.1.2*01(@1,date youngest value:,YYMMdd,Date_youngest,0`
the result is as desired
`221101000000`.
Accessing the decode line (my line is No. 5) with
`dat=sml[5]`
then the result is
`201200992256`
Somewhere, apparently, full double precision is not always used for calculation.

Anyway, thank you and Merry Christmas

Golter

**Reply by gemu2015** [reply]
scripter only supports float values while sml supports only double.
i modified the sources so that you may get a double value as string into scripting.
you may then divide the string into peaces and reconvert to smaller numbers.

smls[M] usually return the string of a meter

now

string = smls[-DL] returns the double number of decode line as string

remark:
the sml driver attached is a complete rework of the last driver with dynamic memory allocation and software settable options instead of #defines. this greatly reduces the RAM consumption especially useful on ESP8266
however it is not yet well tested. i will do a pr next week when i have done more tests.







[Archiv.zip](https://github.com/arendst/Tasmota/files/10296848/Archiv.zip)

**Comment by ciinema** [comment]
@GolterU 

did u get the "current Power" field in your Script. Im asking because we are query the hole Power Meter but cant find the "current power" (1.7.0).

thanks.

**Reply by GolterU** [reply]
Hi,
my smart meter ZE311 (ZPA company) returns a standard data set, which unfortunately also does not contain any instantaneous power (OBIS code 1.7.0). I was in contact with the manufacturer, who wrote to me that it wasn't intended.
I try to get around this by reading the electrical energy meter readings at fixed intervals and calculating the electrical output from them. This method only reaches the limit of the low measurement resolution of the meter readings (1 Wh ≙ 60 Wmin). If you read this every 10 seconds, a minimum power of >= 360 W can be detected - which is very rough. Even at a measuring distance of 1 minute, it is still 60 W.

Best regards
Uwe

**Reply by ciinema** [reply]
Hi Uwe,

Thank you for the Response and the Insides. :)

**Reply by Einwich** [reply]
Is the script for the ZPA ZE311 anywhere available or is it possible to upload it here? I still have an issue to get any answer from the meter.

Thanks a lot
Karsten

**Reply by Einwich** [reply]
I did a lot of beginners mistakes - finally I have a running script. May it is useful for someone else.

`>D
scnt=0
res=0
>B
=>sensor53 r
;=>sensor53 d1
>F
; count 100ms
scnt+=1
switch scnt

;300ms after start: set sml driver to 300 baud and send /?!
case 3
;=>print start sequence
res=sml(1 0 300)
res=sml(1 1 "2F3F210D0A")

;700ms later: Ack and ask for switching to 9600 baud
case 13
;=>print request 9600 Baud
res=sml(1 1 "063035300D0A")

;300ms later: switch sml driver to 9600 baud
case 16
;=>print set driver to 9600 Baud
res=sml(1 0 9600)

;6000ms after start: Restart sequence
case 60

;=>print restart sequence
scnt=0

ends
>M 1
+1,3,o,0,9600,ZE311,1
;1,0.0.0(@#),meter ID ,,Meter_ID,0
;1,0.0.2(@#),Meter number,,Meter_number,0 
1,1.8.0(@1),Total ,KWh,Total,3
1,0.9.1(@#),Time stamp,Uhr,timestamp,0  
#`

**Reply by GolterU** [reply]
Hello,
Here are two script versions that I used. In my experience, there are no difficulties in contacting the mME. However, you are limited to the predefined data set.
The only problem is if you want to switch the data rate from the standard 300 bps to 9600 bps. This seems to fail in an estimated 5% of all attempts. Then the meter remains at 300 bps and the subsequent data reading by the script returns nonsensical data.  I suspect the cause in the Tasmota firmware - some serial level oscillograms look dubious.
I use an ESP-01S with Tasmota 12.4.0 to read two e-meters at the same time. This means that all available ports on the ESP are utilized.
Here is my current script (**without** bitrate switching). I'll leave the comments in, but they're not included in the running version:

```
;Script mit Verwendung des special commands M,2=d x y und ausschließlichem Betrieb der Zählerabfragen mit 300 bps.
;Script enthält Kommentare und Debug-Konsolenausgaben. Entfernen mit Suchen/Ersetzen und regulärem Ausdruck [\x09]*;.* bzw. [\x20]*print.*
;Editieren in VSC empfiehlt sich mit Sprachmodus ini-Dateien -> schön eingefärbter Text
;Stand: 10.10.23

;Leistungsvariablen mit ersten Index 1 (was früher 1-min-Abfragerhythmus bedeutete) beziehen sich jetzt auf 3 min Zeitabstand
>D
res=0
scnt=0
;Die Variablen aus den Meter metrics (Dekodierzeilen) P1EG usw. müssen nicht deklariert werden

>B
->sensor53 r
print Script vom 10.10.2023 22:30, gestartet: %tstamp%\n\r	;Konsolenmarkierung Scriptstart

>M 2	;Beide Zähler auswerten
; Meter Definition Zähler EG (ZNr<n>, RX-GPIO, type OBIS ASCII, flag, bitrate, JSON-Prefix, TX-GPIO, Periode in ms, Request Message als Sendetelegramm)
+1,3,o,0,300,EltEG,1,1630,2F3F210D0A	;Wiederholung der Zählerabfrage aller 163 s (s. Script-Versionen v. heute)
+2,0,o,0,300,EltOG,2,130,2F3F210D0A	;dito
;Textlänge bei M,h<text> auf 26 Zeichen begrenzt! Vermutlich auch bei den anderen M,xxx-Befehlen
;Textlänge Label (nach ..@1,) auf 22 Zeichen begrenzt!
;Maßeinheit max. 7 Z. lang!
1,=hEG, aktuelle Daten
;Meter Metrics (ZNr<n>, Dekodierstring abgeschlossen mit ( dann @ als Dekodierendezeichen. Dann Zahl als Divisor bzw. bei Stringdekodierung #) gefolgt von einem optionalen Offset, Bezeichner für Web-UI, Maßeinheit, MQTT-Bezeichner, Anzahl Dez.stellen). Letzteres bestimmt auch, ob die Daten sofort (dann 16 addieren) oder nur jede Teleperiode gesendet werden.
1,0.0.0(@#),Zählernummer,,Meter_number,0
1,1.8.0(@1,Zählerstand gesamt:,kWh,E1,3
;Leistungsberechnung durch Differenzbildung mit special command. Dekodierter Zählerstand in kWh aus vorangehender Zeile entspricht der Leistung in Watt, wenn jeweils die Parameter nach dem @ 1 sind.
;P[W] =  ΔE[kWh]*1000*3600/Δt[s] Vergleiche "ZPA-ZE311_Script_Versionen.txt" Eintrag vom 10.07.23
1,=d 2 180 @1,3-min-Leistung ±20 W:,W,P1EG,0
1,=d 2 900 @1,15-min-Leistung ±4 W:,W,P15EG,0
1,=hEG, historische Daten
1,0.1.2*01(@1000000,jüngstes Datum:,YYMMDD,Date1_latest,0
1,1.8.0*01(@1,jüngster Zählerstand:,kWh,Consumpt1_latest,3
1,0.1.2*25(@1000000,ältestes Datum:,YYMMDD,Date1_oldest,0
1,1.8.0*25(@1,ältester Zählerstand:,kWh,Consumpt1_oldest,3
1,=h<hr>
1,=hOG, aktuelle Daten
2,0.0.0(@#),Zählernummer,,Meter_number,0
2,1.8.0(@1,Zählerstand gesamt:,kWh,E2,3
2,=d 10 180 @1,3-min-Leistung ±20 W:,W,P1OG,0
2,=d 10 900 @1,15-min-Leistung ±4 W:,W,P15OG,0
1,=hOG, historische Daten
2,0.1.2*01(@1000000,jüngstes Datum:,YYMMDD,Date2_latest,0
2,1.8.0*01(@1,jüngster Zählerstand:,kWh,Consumpt2_latest,3
2,0.1.2*25(@1000000,ältestes Datum:,YYMMDD,Date2_oldest,0
2,1.8.0*25(@1,ältester Zählerstand:,kWh,Consumpt2_oldest,3
#	;Endekennzeichen M-Sektion
```

A script **with** bitrate switching, but with the problem mentioned above:

```
>D
res=0
scnt=0
P1EG=0
P1OG=0
P15EG=0
P15OG=0

>B
->sensor53 r
print Script vom 16.07.2023 13:44, gestartet: %tstamp%\n\r

>F
scnt+=1

switch scnt

case 6
res=sml(1 0 300)
res=sml(1 1 "2F3F210D0A")
res=sml(2 0 300)
res=sml(2 1 "2F3F210D0A")

case 18
res=sml(1 1 "063035300D0A")
res=sml(2 1 "063035300D0A")
res=sml(1 0 9600)
res=sml(2 0 9600)

case 100
scnt=0

ends

>M 2
+1,3,o,0,9600,EltEG,1
+2,0,o,0,9600,EltOG,2
1,=hEG, aktuelle Daten
1,0.0.0(@#),Zählernummer,,Meter_number,0
1,1.8.0(@1,Zählerstand gesamt:,kWh,E1,3
1,=d 2 60 @1,1-min-Leistung ±60 W:,W,P1EG,0
1,=d 2 900 @1,15-min-Leistung ±4 W:,W,P15EG,0
1,=hEG, historische Daten
1,0.1.2*01(@1000000,jüngstes Datum:,YYMMDD,Date1_latest,0
1,1.8.0*01(@1,jüngster Zählerstand:,kWh,Consumpt1_latest,3
1,0.1.2*25(@1000000,ältestes Datum:,YYMMDD,Date1_oldest,0
1,1.8.0*25(@1,ältester Zählerstand:,kWh,Consumpt1_oldest,3
1,=h<hr>
1,=hOG, aktuelle Daten
2,0.0.0(@#),Zählernummer,,Meter_number,0
2,1.8.0(@1,Zählerstand gesamt:,kWh,E2,3
2,=d 10 60 @1,1-min-Leistung ±60 W:,W,P1OG,0
2,=d 10 900 @1,15-min-Leistung ±4 W:,W,P15OG,0
1,=hOG, historische Daten
2,0.1.2*01(@1000000,jüngstes Datum:,YYMMDD,Date2_latest,0
2,1.8.0*01(@1,jüngster Zählerstand:,kWh,Consumpt2_latest,3
2,0.1.2*25(@1000000,ältestes Datum:,YYMMDD,Date2_oldest,0
2,1.8.0*25(@1,ältester Zählerstand:,kWh,Consumpt2_oldest,3
#
```
Good luck!
Uwe

**Reply by Einwich** [reply]
Thanks a lot. I used your power calculations and it works as expected.

Karsten

**Reply by schokominzpfefferlade** [reply]
Hi,

I just don't get any values from the meter. Tried with script from Einwich:

![grafik](https://github.com/arendst/Tasmota/assets/60002031/c7fd2e87-2eed-4464-8634-776ee97e4e3b)

with scripts from GolterU it´s the same:
![grafik](https://github.com/arendst/Tasmota/assets/60002031/24f2b0ea-11b0-4561-80bc-506857488229)

What am I doing wrong?
Does the meter have to be unlocked by a PIN? I asked my electricity provider and they said "no".
The reader is correctly placed on the IR interface.
Any ideas?

**Reply by Noschvie** [reply]
Have you tried the debug mode?

**Reply by schokominzpfefferlade** [reply]
you mean the "sensor53 d1" command? i tried but it didn't make a difference

**Reply by YoungJules** [reply]
That command turns on dumping the raw SML data to the console...(sorry for the late reply!)
